### DISCUSSION TITLE
New package structure

### DISCUSSION THREAD

**Discussion by schuenke** [imr-framework/pypulseq#123] [General]
Currently, PyPulseq uses a package structure without dedicated ```src/``` folder and most files are located in the ```pypulseq/``` subfolder directly. As we are planning to move away from SimpleNamespaces to a more objected-oriented implementation (#95), it makes sense to adjust the package structure in my opinion.

I would suggest a structure including a ```src/``` folder as officially recommended by python.org (https://packaging.python.org/en/latest/tutorials/packaging-projects/) and dedicated subfolders for the individual submodules.

Such a structure could look like this:
```
pypulseq/
├── LICENSE
├── pyproject.toml
├── README.md
├── src/
│   └── pypulseq/
│       ├── __init__.py
│       ├── sequence
│            ├── __init__.py
│            ├── sequence.py
│            ├── write_seq.py
│             ⋮
│       ├── grad
│            ├── __init__.py
│            ├── grad.py
│            ├── make_trapezoid.py
│             ⋮
│       ├── rf
│            ├── __init__.py
│            ├── rf.py
│            ├── make_block_pulse.py
│            ├── make_gauss_pulse.py
│             ⋮
│       ├── ...
└── tests/
```

Using the `__init__.py` files we could ensure that all current imports statements work and thus ensure compatibility for all user scripts. 

Maybe we can use this thread to discuss advantages and disadvantages of different structures and find a consensus regarding subfolders/subpackages/submodules. 

Some open questions would be:

1. Where to put all "block" related definitions/functions?
2. Where to put functions that are shared between submodules (e.g. `compress_shape` is used for rf pulses and arbitrary gradients, `calculate_duration` is used for all event types (RF, ADC, GRAD, DELAY)
3. Should we use the same subfolder structure in the `tests/` folder?

Looking forward to hear your opinions and please feel free to add any other ideas/questions that come to your mind!

**Comment by sravan953** [comment]
@schuenke , thank you for opening this discussion thread! I agree with almost everything you've stated, we are on the same page.

To answer the points you put up, I think the block-related functions should go into a `block` submodule, and the other common functionalities into a `utils` submodule. And yes, let us use the same structure for `tests`.

**Reply by schmidtijoe** [reply]
I would also vote for a block submodule. In my own hacky implementation i called it kernel and its nice to have some features for block creation like some methods for building slice selective excitation or a specific readout that already sets phase encode and read gradients along with the adc. makes building sequences much more convenient.

**Comment by schote** [comment]
Thanks @schuenke, I like the structure you proposed. I agree to @sravan953 and would use a `tools/` or `utils/` folder for shared functionalities. I'm not sure how many tests are needed per submodule. Do you think we even need a subfolder structure for `tests/`?

**Comment by schuenke** [comment]
closed with #218
